compiler/resolution/implementForallIntents.cpp

```diff
@@ -1778,38 +1878,221 @@ static bool isFsIndexVar(ForallStmt* fs, Symbol* sym)
+    // fiVar should have been resolved already
+    INT_ASSERT(eltType != dtUnknown && eltType != dtAny);
+
+    if (eltType->symbol->hasFlag(FLAG_HAS_RUNTIME_TYPE)) {
+      // arg = typeof(fiVar)
+      VarSymbol* typeofTemp = newTemp("chpl__typeofET");
+      typeofTemp->addFlag(FLAG_TYPE_VARIABLE);
+      hld->insertAtTail(new DefExpr(typeofTemp));
+      hld->insertAtTail("'move'(%S,'typeof'(%S))", typeofTemp, fiVarSym);
+      eltTypeArg = new SymExpr(typeofTemp);
+    } else {
+      // Forego typeof(), use the type directly.
+      eltTypeArg = new SymExpr(eltType->symbol);
+    }
+  }
+
+  hld->insertAtTail("'move'(%S, 'new'(%S,%E))", globalOp, riTypeSym,
+                    new NamedExpr("eltType", eltTypeArg));
+  fs->insertAfter("chpl__delete(%S)", globalOp);
+  insertGenerateAfter(fs, fiVarSym, globalOp);
+
+  resolveBlockStmt(hld);
+  hld->flattenAndRemove();
+
+  // Todo: this replace() is somewhat expensive.
+  // Can instead we update fs->riSpecs[i] aka 'riSpec' in-place?
+  origRiSpec->replace(new SymExpr(globalOp));
+  
+  return globalOp;
+}
+
+static Symbol* handleRISpec(ForallStmt* fs, Expr* fiVar, Expr* riSpec) {
+  if (!fs)
+    // old style
+    return toSymExpr(riSpec)->symbol();
+
+  Symbol* result = NULL;
+  Symbol* fiVarSym = toSymExpr(fiVar)->symbol();
+  SET_LINENO(riSpec);
+
+  if (SymExpr* riSE = toSymExpr(riSpec)) {
+    Symbol* riSym = riSE->symbol();
+
+    if (TypeSymbol* riTypeSym = toTypeSymbol(riSym)) {
+      result = setupRiGlobalOp(fs, fiVarSym, riSpec, riTypeSym, NULL);
+
+    } else {
+      if (isArgSymbol(riSpec))
+        // This could be confused with a tiMark. Need to handle it.
+        INT_FATAL(fs, "not implemented");
+
+      // The user will manage allocation and deallocation of 'riSym'.
+      // So we only need to add a call generate().
+      insertGenerateAfter(fs, fiVarSym, riSym);
+      result = riSym;
+    }
+
+  } else if (CallExpr* riSpecCall = toCallExpr(riSpec)) {
+    SymExpr* riTypeSE = toSymExpr(riSpecCall->baseExpr);
+    INT_ASSERT(riTypeSE); // need to handle baseExpr anyway
+    TypeSymbol* riTypeSym = toTypeSymbol(riTypeSE->symbol());
+    INT_ASSERT(riTypeSym); // non-types should be handled elsewhere
+
+    if (riSpecCall->numActuals() != 1)
+      USR_FATAL(riSpecCall, "for a reduce intent, the 'reduce' keyword must be preceded by the reduction operator or the name of the reduction class with the single optional argument indicating the type of the reduction input");
+
+    riTypeSE->remove();
+    Expr* eltTypeArg = riSpecCall->get(1)->remove();
+    result = setupRiGlobalOp(fs, fiVarSym, riSpec, riTypeSym, eltTypeArg);
+
+  } else {
+    // What else can this be?
+    INT_FATAL(fs, "not implemented");
+  }
+
+  return result;
+}
+
+// Mark the variables listed in 'with' clauses, if any, with tiMark markers.
+// Same as markOuterVarsWithIntents() in createTaskFunctions.cpp,
+// except uses ForallIntents.
```

Comment:
We need to convert other forall loops to the new representation
to retire the old implementation.
I do not have a specific date set for this work.
This PR does not address it.